django/applications/catmaid/migrations/0006_add_history_table_support.py

```diff
@@ -0,0 +1,501 @@
+                'CREATE TABLE IF NOT EXISTS %I (LIKE %s) INHERITS (%I)',
+                history_table_name,live_table_name, parent_history_table_name
+            );
+        ELSE
+            -- Create a regular history table without inheritance, either
+            -- because no parent is available or no parent check was performed.
+            EXECUTE format(
+                'CREATE TABLE IF NOT EXISTS %I (LIKE %s)',
+                history_table_name,live_table_name
+            );
+        END IF;
+
+        -- Make all history columns (except the later added sys_period column
+        -- default to NULL
+        FOR column_info IN
+            SELECT c.column_name
+            FROM information_schema.columns c
+            WHERE c.table_schema NOT IN ('information_schema', 'pg_catalog')
+            AND c.table_name = live_table_name::text
+            AND c.column_name <> 'sys_period'
+        LOOP
+            -- Drop NOT NULL constraints and add default
+            EXECUTE format(
+                'ALTER TABLE %I ALTER COLUMN %I DROP NOT NULL',
+                history_table_name, column_info.column_name);
+            EXECUTE format(
+                'ALTER TABLE %I ALTER COLUMN %I SET DEFAULT NULL',
+                history_table_name, column_info.column_name);
+        END LOOP;
+
+        -- Add a system time column to the history table, named sys_period, if
+        -- it doesn't exist already (which can happen due to table inheritence.
+        IF NOT EXISTS(SELECT column_name
+                      FROM information_schema.columns
+                      WHERE table_schema = 'public'
+                      AND table_name = history_table_name
+                      AND column_name = 'sys_period') THEN
+            EXECUTE format(
+                'ALTER TABLE %I ADD COLUMN sys_period tstzrange
+                NOT NULL DEFAULT tstzrange(current_timestamp, null)',
+                history_table_name
+            );
+        END IF;
+
+        -- Add a transaction reference to the history table, named
+        -- exec_transaction_id, if it doesn't exist already (which can
+        -- happen due to table inheritence. Together with the lower part
+        -- of the sys_period range, the transaction ID is uniqie.
+        IF NOT EXISTS(SELECT column_name
+                      FROM information_schema.columns
+                      WHERE table_schema = 'public'
+                      AND table_name = history_table_name
+                      AND column_name = 'exec_transaction_id') THEN
+            EXECUTE format(
+                'ALTER TABLE %I ADD COLUMN exec_transaction_id bigint
+                NOT NULL DEFAULT txid_current()',
+                history_table_name
+            );
+        END IF;
+
+        -- Create sys_period (validity period) index for the new history
+        -- table. This is needed to quickly query older versions of an
+        -- entity.
+        -- TODO: Maybe also needs an id index?
```

Comment:
Yeah `BRIN` indexes might be useful here: edition time will always get newer and columns with the same ID are likely to be accessed together. But let's wait until we understand the access patterns better that we are going to have.